### PR TITLE
Uniswap V3 Split Router Callee

### PR DESCRIPTION
**Pull request by KirillDogadin-std**
This PR adds new Uniswap Split Route callee contract and its tests.

Notes:

- The contract was developed using [foundry](https://book.getfoundry.sh/)
- Since Split route also requires data provided by the [Uniswap AlphaRouter](https://github.com/Uniswap/smart-order-router) to operate we:
  - Hardcoded example data to the test contract
  - Provided node.js script to fetch fresh data and then pass it to the same test via env variables

### Functional testing

1. Create `.env` file (in the root of the project) containing valid rpc url under `FOUNDRY_ETH_RPC_URL` name
2. Run tests
      - Using hardcoded alpha router data: `forge test --match testTakeLinkUniswapSplitProfit`
      - Using freshly fetched alpha router data: `(cd scripts/uniswap-split-route-callee && npm ci && node index.js)`

### CODE REVIEW

File: src/UniswapV3SplitRouteCallee.sol [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+interface UniV3RouterLike {
+    function multicall(uint256 deadline, bytes[] calldata data) external payable returns (bytes[] memory results);
+}
+
+contract UniswapV3SplitCallee {
+    UniV3RouterLike public uniswapV3Router;
+    DaiJoinLike     public daiJoin;
+    TokenLike       public dai;
```
Comment (rockyfour) [inline src/UniswapV3SplitRouteCallee.sol:48]: These should be immutable.

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:17]: I think this file has gone some auto formatting vs the 1inch test (e.g changing " -> ', spaces alignments etc).
Can we maybe revert that and if we think formatting fixes are needed it should be a separate PR that covers more than this file?

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+pragma solidity =0.6.12;
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:18]: Why the change from `>=` to `=`?

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+}
+
+struct UniswapV3ExactInputParams {
+    bytes path;
+    address recipient;
+    uint256 amountIn;
+    uint256 amountOutMinimum;
+}
+
+interface WethAbstract is GemAbstract {
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:45]: All weth related stuff are unused, please remove as in the 1inch test.

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+    bytes   uniswapTxDataNoProfit;
+
+    function getPermissions() private {
+        hevm.store(dogAddr, keccak256(abi.encode(address(this), uint256(0))), bytes32(uint256(1)));
+        hevm.store(vatAddr, keccak256(abi.encode(address(this), uint256(0))), bytes32(uint256(1)));
+        hevm.store(linkPipAddr, keccak256(abi.encode(address(this), uint256(0))), bytes32(uint256(1)));
+        linkPip.kiss(address(this));
+    }
+
+    function setEnvVars() private {
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:163]: Please add an explanation as in 1inch test.

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+        /* hevm.rollFork(uniswapBlock); */
+    }
+
+    function setUp() public {
+        setAddresses();
+        setInterfaces();
+        setEnvVars();
+        ali = new VaultHolder(vat);
+        aliAddr = address(ali);
+        dan = new UniswapV3SplitCallee(address(0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45), daiJoinAddr);
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:185]: Ideally, the router address would be defined as a constant that would be used here.

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+        for (uint256 i = 0; i < calls.length; i++) {
+            (bytes memory signature, bytes memory callWithTrimmedHash) = this.trimFunctionHash(calls[i]);
+            UniswapV3ExactInputParams memory mCalldata = abi.decode(callWithTrimmedHash, (UniswapV3ExactInputParams));
+            UniswapV3ExactInputParams memory modifiedParams = UniswapV3ExactInputParams({
+                path: mCalldata.path,
+                recipient: address(dan),
+                amountIn: mCalldata.amountIn,
+                amountOutMinimum: mCalldata.amountOutMinimum
+            });
+            transformedCalls[i] = this.buildCalldata(signature, abi.encode(modifiedParams));
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:278]: Is there a reason for making it an external call?

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+        uint256 kicksPost = linkClip.kicks();
+        assertEq(auctionId, kicksPost);
+        assertEq(kicksPost, kicksPre + 1);
+        (,, uint256 lot, address usr, uint96 tic,) = linkClip.sales(auctionId);
+        assertEq(usr, aliAddr);
+        assertEq(lot, amountLink);
+        assertEq(tic, block.timestamp);
+    }
+
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:301]: Are these math functions needed?

File: scripts/uniswap-split-route-callee/index.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
+import path from 'path';
+import url from 'url';
+import * as dotenv from 'dotenv';
+import { spawn } from 'child_process';
+import { AlphaRouter, ChainId, SwapType } from '@uniswap/smart-order-router';
+import { CurrencyAmount, Percent, Token, TradeType } from '@uniswap/sdk-core';
+import { ethers } from 'ethers';
+
```
Comment (rockyfour) [inline scripts/uniswap-split-route-callee/index.js:8]: Can we try and make this script as similar as possible to it's 1inch counterpart?
For example the 1inch has 
Ideally the scripts would be almost identical in a diff except some well contained parts.

Would be good if things like this are the same when diffing - order of imports, structure and name of functions, rpc related definitions/assignments, link input amounts...

**Review by rockyfour — COMMENTED** [review]
Looks good in general.
See some comments, mostly targeting the script and test to be as similar as possible to the 1inch counterpart.

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 8 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol]: removed

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 6 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol]: extracted

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 5 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol:156]: added

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 4 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol]: Weth stuff removed.

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 3 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol]: Sausage fingers hitting backspace when they're not supposed to.
Retured to `>=`

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 2 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol:17]: I've tried to pinpoint the changes with the eyes, only noticed the quote differences.

Do you have any formatting process that you perform before PRs usually? e.g. some plugin in the VSCode that i could also run?

Or if you don't i would appreciate if you could enumerate some other issues you see with the formatting.

File: src/UniswapV3SplitRouteCallee.sol (hunk 1 above)
Comment (KirillDogadin-std) [reply src/UniswapV3SplitRouteCallee.sol]: done

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 7 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol]: Adjusted.

File: scripts/uniswap-split-route-callee/index.js (hunk 9 above)
Comment (KirillDogadin-std) [reply scripts/uniswap-split-route-callee/index.js:8]: So far i have adjusted the import order and the usage of the env variable.

For other adjustments (if they are necessary) i would have the following questions:

  1. Are you implying that i should also adjust the 1inch script to achieve the maximal similarity?
  2. By similarities "structure of functions" do you mean "as many similar lines as possible"? something else?
  3. Since from my perspective the processes of setting up the tests via scripts are quite different, i would also not expect the scripts to be similar. So in order to make a meaningful adjustments, i would appreciate if you could select 1 particular instance of something that should be adjusted, describe (on very general level) what the adjustment would look like and what the adjustment improves particularly (or 1 main point). Then i will take this with me and propagate the approach on the whole file.

**Comment by rockyfour** [comment]
@KirillDogadin-std @valiafetisov in the 1Inch integration the actual collateral slice amount and owed debt to be paid back are re-encoded in the callee itself - 
https://github.com/makerdao/exchange-callees/blob/1f9ba7b0e5d202774212c8ccfb9b0ac929ee9124/src/OneInchCallee.sol#L131-L132

Here this is not done.
So in practice, the `slice` parameter, which is calculated in the clipper is ignored, and you might end up not swapping all the collateral you got from the clipper.

Was the approach from 1Inch changed intentionally?

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 2 above)
Comment (rockyfour) [reply src/test/UniswapV3SplitRouteCallee.t.sol:17]: Mmm we don't have an automatic formating process (but it looks like maybe you do).
I guess some of it is not only auto-formatting, but here is some stuff that is different than the 1Inch test (and makes the review process harder):

* import from `src` and not `..`
* having named vars in `function store`
* changing to `vat_` and not `vat` in VaultHolder constructor
* 1E18 => 1e18
* spacing in definition of GemAbstract, VatAbstract etc..
* spacing in definitions of VaultHolder and addresses below it
* spaces in `getLinkPriceRay`
* spaces in `frobMax`
* spaces in `testDrip`
* spaces in `testBarkLink`
* order of var definitions in `testTakeLinkUniswapSplitProfit`, `testTakeLinkUniswapSplitProfit`..

I guess some of these changes are good, but best to do them separately to ease the review.

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 10]
```diff
@@ -0,0 +1,359 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2023 Maker Ecosystem Growth Holdings, INC.
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:2]: This is not needed anymore as far as I'm aware.

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+        hevm.store(vatAddr, keccak256(abi.encode(address(this), uint256(0))), bytes32(uint256(1)));
+        hevm.store(linkPipAddr, keccak256(abi.encode(address(this), uint256(0))), bytes32(uint256(1)));
+        linkPip.kiss(address(this));
+    }
+
+    function setEnvVars() private {
+        // Execute this test with fresh data via
+        // `(cd scripts/uniswap-split-route-callee && npm ci && node index.js)`
+        // check the script for more details on how to use universal router
+        UniswapV3Router2 = hevm.envOr('UniswapV3Router2', 0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45);
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:160]: Should use the new constant as the right parameter, left param env var style should be same as `ONE_INCH_ROUTER`.

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+        uniswapTxDataNoProfit = hevm.envOr(
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:161]: How was the no profit vs profit amount generated?
Couldn't get it from looking at the script as well.

I guess I also don't really get why the multicall data should be different and separated to cases, as there is only one best uniswap price at a given time (and the min profit amount is enforced in the callee).
Maybe I'm missing something.

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+        bytes memory data = abi.encode(danAddr, linkJoinAddr, minProfit, address(0), reencodedAutoRouterData(txData));
+        linkClip.take(auctionId, amt, max, danAddr, data);
+    }
+
+    function createLinkAuction()
+        private
+        returns (uint256 auctionId, uint256 amountLinkWad, uint256 auctionPrice, uint256 auctionDebt)
+    {
+        (,,,, uint256 dustRad) = vat.ilks(linkName);
+        amountLinkWad = (dustRad / getLinkPriceRay()) * 2;
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:309]: Can you please also fix this (`amountLinkRay` => `amountLinkWad`) in the 1Inch callee test?

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 12 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol:161]: > How was the no profit vs profit amount generated?

via using the different `linkAmountToSwap` values.

- if we derive this amount from 1.5*dust (in link equivalent): the this will be "no profit" for the coded test scenario
- if we derive this from 2*dust: this is the case where it has profit.

i.e. script function `getLinkAmountToSwap(2, 1)` yields us the case with sufficient profit, `getLinkAmountToSwap(3, 2)` - with no profit.

for "why this happens this way? why can't we use same link amounts?" see the bottom-most section 

> as there is only one best uniswap price at a given time

that is true, but for different amounts of input currency the return of the `AlphaRouter` can potentially be different (e.g. in distribution if currency among routes, routes themselves) as of [the description](https://uniswap.org/blog/auto-router)

___

Why different link amounts?

during development of the no-profit test the exchange with same calldata always resulted into profitable case. So this required to fine-tune the swapped amounts.
With same calldata (same link amount) as in profit case taking the auction as soon as it stops being non-profitable still results into profit. Making the link amount smaller resolved this problem.

**Comment by KirillDogadin-std** [comment]
https://github.com/makerdao/exchange-callees/pull/35#issuecomment-1457771576

> Was the approach from 1Inch changed intentionally?

the calldata that is generated by AlphaRouter has a non-trivial structure and is supposed to be (expected to be) highly optimised for the specific input amount. This means we should choose either or:
  - do repackage the data, write several not-easy to understand and implement operations to unpack the data (of non-fixed size) and validate that `slice` amount matches the cumulative sum of routes' input. 
  - assume that the wallet that calls the contract is responsible for supplying the matching (matching to the Universal Router calldata) `slice` value, keep the contract simple and less expensive.

The latter was preferred according to https://github.com/makerdao/exchange-callees/pull/34#discussion_r1066712189

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 11 above)
Comment (rockyfour) [reply src/test/UniswapV3SplitRouteCallee.t.sol]: mm I think you missed my suggestion to use `uniV3Router` constant instead of the explicit address here.

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 13 above)
Comment (rockyfour) [reply src/test/UniswapV3SplitRouteCallee.t.sol:399]: There are more places in the callees test, for example everywhere we call `createLinkAuction`.
Suggest to search for `amountLinkRay` and validate each occurrence.

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 2 above)
Comment (rockyfour) [reply src/test/UniswapV3SplitRouteCallee.t.sol:16]: mm I don't think this has been handled.
Just adding a diff of the top of this file and the 1inch equivalent one to show what I mean:

image

File: scripts/uniswap-split-route-callee/index.js (hunk 9 above)
Comment (rockyfour) [reply scripts/uniswap-split-route-callee/index.js:8]: 1. If you determine that something in the 1inch script needs to change due to the way it's used in the Uniswap script then yes. If it's just a style issue or a non-mandatory improvement then at this point I would just do it like the 1Inch script (and possibly improve both in a separate PR).

2. I mean that when doing a visual diff it will be very clear what's different between 1Inch and Uniswap router. See (3) for an example. Eventually, the scripts do (or should do) the same thing, just using different platforms.

3. An example is the new script has a `getProvider` function while the 1Inch script does not. I understand you need the provider in more places in the new script so in this case I'd advise aligning the 1Inch script (and making sure to re-test it ofc).
Another example is the new `getLinkAmountToSwap` function, we didn't have that before. It might be currently needed since you differentiate the profit vs profit cases according to the input amount, but it should actually be done according to when the take is done during the auction (see below). If we use a fix as in the 1Inch then we can make that part similar as well.

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 12 above)
Comment (rockyfour) [reply src/test/UniswapV3SplitRouteCallee.t.sol:209]: I think if it makes your life much easier you can remove the no profit test.

That's maybe actually cleaner than purposely swapping only part of the taken auctioned collateral so there will be not much dai profit. That part is really confusing as it's different than all the other tests in this repo.

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 13 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol:315]: adjusted

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 12 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol]: removed the no-profit test

File: scripts/uniswap-split-route-callee/index.js (hunk 9 above)
Comment (KirillDogadin-std) [reply scripts/uniswap-split-route-callee/index.js:8]: - reshuffled the function structure a bit
- the amount to swap is now hardcoded.

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 2 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol:16]: oh, really sorry about this one 😓 , i probably pressed "Save" instead of "Save without formatting" at some point.

Should be either good or better now :)

File: src/test/UniswapV3SplitRouteCallee.t.sol (hunk 11 above)
Comment (KirillDogadin-std) [reply src/test/UniswapV3SplitRouteCallee.t.sol]: addressed

**Comment by rockyfour** [comment]
Almost there.
I do see this failure (let's ignore other instabilities for the purpose of merging this):
[FAIL. Reason: STF] testTakeLinkUniswapSplitProfit() (gas: 8260641125390436681)
this was on block 16941906

Will approve after that test is fixed.

**Comment by KirillDogadin-std** [comment]
So this test fails sometimes due to the link price fluctuation (solidity test determines the link amount used in the test based on `dust` and link price at a block height). With the hardcoded amount there's a risk that eventually the test will fail either with STF or InsufficientProfit. To avoid this, the initial dynamic computation of the link amount was introduced. 

I've adjusted the amount of link that the script provides, it works out now, but i am not even sure that when you will take a look at this next time it will pass.

If you have a preference on how to address this (if this should be addressed) please let me know.

File: src/test/UniswapV3SplitRouteCallee.t.sol [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+        // check the script for more details on how to use universal router
+        uniswapV3Router2 = hevm.envOr('UNISWAP_V3_ROUTER', uniV3Router);
+        uniswapTxDataProfit = hevm.envOr('UNISWAP_TX_DATA_PROFIT', hex'0000000000000000000000000000000000000000000000000000000063c7bba90000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000124b858183f00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000dead0000000000000000000000000000000000000000000000eb50c6636ac5f312f00000000000000000000000000000000000000000000005a1d6fc842061b484280000000000000000000000000000000000000000000000000000000000000059514910771af9ca656af840dff83e8264ecf986ca000bb8c02aaa39b223fe8d0a0e5c4f27ead9083c756cc20001f4a0b86991c6218b36c1d19d4a2e9eb0ce3606eb480000646b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000000000000000000000000000');
```
Comment (rockyfour) [inline src/test/UniswapV3SplitRouteCallee.t.sol:159]:
```suggestion
        uniswapTxDataProfit = hevm.envOr('UNISWAP_TX_DATA_PROFIT', hex'00000000000000000000000000000000000000000000000000000000660a582d0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000124b858183f00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000dead0000000000000000000000000000000000000000000000c2f7b1f2632c53ee5c00000000000000000000000000000000000000000000052e05ebb83d0073467a0000000000000000000000000000000000000000000000000000000000000059514910771af9ca656af840dff83e8264ecf986ca000bb8c02aaa39b223fe8d0a0e5c4f27ead9083c756cc20001f4a0b86991c6218b36c1d19d4a2e9eb0ce3606eb480000646b175474e89094c44da98b954eedeac495271d0f0000000000000000000000000000000000000000000000000000000000000000000000');
```

**Comment by rockyfour** [comment]
> I've adjusted the amount of link that the script provides, it works out now, but i am not even sure that when you will take a look at this next time it will pass.

Ok, you didn't update the test itself with the new txData, so I added a suggestion for it above.

With that change the test passes (block 16959704).
Note that I didn't go deeply into the new amount you set and whether it could be made more robust. You can consider looking at that further and maybe making changes in a separate PR if needed.

Please commit the above suggestion when you can, then I will approve and merge the PR.

**Comment by KirillDogadin-std** [comment]
Committed, thanks for including suggestion.

**Review by rockyfour — APPROVED** [review]
LGTM
